server/lib/amplitude.js

```diff
@@ -0,0 +1,245 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// This module contains mappings from content server event names to
+// amplitude event definitions. The intention is for the returned
+// `receiveEvent` function to be invoked for every event and the
+// mappings determine which of those will be transformed into an
+// amplitude event.
+//
+// You can read more about the amplitude event
+// structure here:
+//
+// https://amplitude.zendesk.com/hc/en-us/articles/204771828-HTTP-API
+//
+// You can see the event taxonomy here:
+//
+// https://docs.google.com/spreadsheets/d/1G_8OJGOxeWXdGJ1Ugmykk33Zsl-qAQL05CONSeD4Uz4
+
+/* eslint-disable sorting/sort-object-props, camelcase */
+
+'use strict';
+
+const APP_VERSION = /^[0-9]+\.([0-9]+)\./.exec(require('../../package.json').version)[1];
+
+const GROUPS = {
+  email: 'fxa_email',
+  login: 'fxa_login',
+  registration: 'fxa_reg',
+  settings: 'fxa_pref',
+  sms: 'fxa_sms'
+};
+
+const ENGAGE_SUBMIT_EVENT_GROUPS = {
+  'force-auth': GROUPS.login,
+  signin: GROUPS.login,
+  signup: GROUPS.registration
+};
+
+const VIEW_EVENT_GROUPS = {
+  'force-auth': GROUPS.login,
+  settings: GROUPS.settings,
+  signin: GROUPS.login,
+  signup: GROUPS.registration,
+  sms: GROUPS.sms
+};
+
+const EMAIL_TYPES = {
+  'complete-reset-password': 'reset_password',
+  'complete-signin': 'login',
+  'verify-email': 'registration'
+};
+
+const NEWSLETTER_STATES = {
+  optIn: 'subscribed',
+  optOut: 'unsubscribed',
+};
+
+const EVENTS = {
+  'flow.reset-password.submit': {
+    group: GROUPS.login,
+    event: 'forgot_submit'
+  },
+  'settings.change-password.success': {
+    group: GROUPS.settings,
+    event: 'password'
+  },
+  'settings.clients.disconnect.submit': {
+    group: GROUPS.settings,
+    event: 'disconnect_device'
+  },
+  'settings.signout.success': {
+    group: GROUPS.settings,
+    event: 'logout'
+  },
+};
+
+const FUZZY_EVENTS = new Map([
+  [ /^flow\.([\w-]+)\.engage$/, {
+    isDynamicGroup: true,
+    group: eventCategory => ENGAGE_SUBMIT_EVENT_GROUPS[eventCategory],
+    event: 'engage'
+  } ],
+  [ /^flow\.[\w-]+\.forgot-password$/, {
```

Comment:
How come some of these have capturing groups while others do not?

Reply:
Can you document the format so that future devs know what to add?

Reply:
The format of what, the regex? There's no format as such, because there is no single format for our events. Each regex is to match a different event, or set of events.

Reply:
> There's no format as such, because there is no single format for our events. 

What I'm getting at is below [match[1]](https://github.com/mozilla/fxa-content-server/pull/5412/files/f2a1c37fb3c568a2f3c9b36d1313d61de64564cc#diff-2a83e43540a70e8aa80daa6c26a3de6dR148) is referenced. What's that for? Where does the data come from? ISTM it's filled in from the contents captured the capturing parens, but to get that far, I had to read the code several times. Someone who is adding a new rule isn't going to want to read the code to figure out what's going on, they are going to want to read a description of what can be in the regexps.